### PR TITLE
Update flake8 configuration

### PR DESCRIPTION
**Pull request by ewu63**
## Purpose
This PR updates the flake8 configuration:
- added the following plugins and checks
	- flake8-bugbear
	- flake8-comprehensions
	- flake8-docstrings
	- flake8-builtins
	- flake8-rst-docstrings
- updated the flake8 config file to reflect those changes
- switched to using requirements files for pinned versions of python packages

I have tested the config file locally, but we can only test Azure after merging this. I wouldn't be surprised if something broke and I will try to patch it after.

Everything here is up for discussion, I know this is rather intrusive so happy to change/reduce the scope if needed.

### CODE REVIEW

**Comment by A-Gray-94** [comment]
Could you explain what these new flake8 checks will catch and what the reason for switching to requirements files is?

File: azure/azure_style.yaml [hunk 1]
```diff
@@ -52,7 +52,7 @@ jobs:
           versionSpec: "3.8"
       - script: |
           cd ${{ parameters.REPO_NAME }}
-          pip install black==21.6b0
+          pip install -r ../.github/black-requirements.txt
```
Comment (sseraj) [inline azure/azure_style.yaml:55]: black currently doesn't checkout `.github`

Comment (nwu63) [reply azure/azure_style.yaml:55]: Yeah I guess this requirements file is unnecessary, I can change it back.

**Comment by nwu63** [comment]
> Could you explain what these new flake8 checks will catch and what the reason for switching to requirements files is?

- [flake8-bugbear](https://github.com/PyCQA/flake8-bugbear) catches common bad coding practices. This IMO is the most important one out of these plugins (it's also maintained by pyCQA so it's maybe more reputable). See [here](https://github.com/PyCQA/flake8-bugbear#list-of-warnings) for a list of things they catch.
- [flake8-comprehensions](https://github.com/adamchainz/flake8-comprehensions) is for improving dict/set/list comprehensions. Not bugs or anything, just improvement in code quality and making things more pythonic.
- [flake8-builtins](https://github.com/gforcada/flake8-builtins) checks for instances where variable names overshadow a python builtin, such as `set = 10`. In most cases this is probably okay, but still considered kinda dangerous IMO.
- [flake8-docstrings](https://github.com/PyCQA/flake8-docstrings) is another plugin from pyCQA, it adds pydocstyle to flake8 so we can lint docstrings according to a given convention (in this case numpydoc). It will complain if there is missing docstring, or if a parameter is missing or anything like that. __This check is currently disabled because very few of our docstrings are up to standard.__
- [flake8-rst-docstrings](https://github.com/peterjc/flake8-rst-docstrings) is for linting RST compliance with our docstrings, whereas the previous plugin just checks for plaintext. I actually think this is probably not that necessary since the plugin is also very much in alpha/beta status. One related thing we could do instead (and I've talked to @eirikurj about this) is to make sphinx build warnings into errors on RTD instead.

I moved the installation to a requirements file to manage the environment a bit more easily, particularly if we want to pin specific versions of packages (as we do now with `flake8`). I didn't like having those specified in the command line in a YAML file somewhere. I will remove the black requirements file as per the discussion above, but I think the flake8 stuff can probably benefit from this sort of partition.

**Comment by marcomangano** [comment]
>  One related thing we could do instead (and I've talked to @eirikurj about this) is to make sphinx build warnings into errors on RTD instead.

Can you clarify this? We want sphinx to fail if the linting is not RST compliant? Is this to avoid/minimize weird renderings on RTD?

**Comment by nwu63** [comment]
> Can you clarify this? We want sphinx to fail if the linting is not RST compliant? Is this to avoid/minimize weird renderings on RTD?

Not the linting, but when building sphinx docs, sometimes we get a bunch of warnings (such as target not found, section heading inconsistent, etc.) but those are not reported by RTD. If you are not checking the logs you have no idea anything was wrong at all. The idea would be to change the `makefile` for sphinx such that it fails on warnings.

**Comment by marcomangano** [comment]
Oh, I like this, although we have to be careful about which warnings to include, some of them are pretty harmless, others are expected to pop up locally (e.g. links to private repo docs)

**Comment by nwu63** [comment]
> Oh, I like this, although we have to be careful about which warnings to include, some of them are pretty harmless, others are expected to pop up locally (e.g. links to private repo docs)

This is a good point, I tried to mock something up but could not bypass this without leaking the RTD tokens in PRs. Let's shelve this for now, and hopefully people who are editing docs are examining their build logs (either locally or on RTD).

**Comment by marcomangano** [comment]
Alright, so are we converging on using `flake8-bugbear` and `flake8-builtins`? I am also fine with including `flake8-comprehensions`

**Pull request merged**